src/Aspire.Hosting.AWS/Provisioning/Provisioners/AWSProvisioner.cs

```diff
@@ -0,0 +1,167 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Lifecycle;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Aspire.Hosting.AWS.Provisioning;
+
+internal sealed class AWSProvisioner(
+    DistributedApplicationExecutionContext executionContext,
+    IServiceProvider serviceProvider,
+    ResourceNotificationService notificationService,
+    ResourceLoggerService loggerService) : IDistributedApplicationLifecycleHook
+{
+    public async Task BeforeStartAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken = default)
+    {
+        if (executionContext.IsPublishMode)
+        {
+            return;
+        }
+
+        var awsResources = appModel.Resources.OfType<IAWSResource>().ToList();
+        if (awsResources.Count == 0)
+        {
+            return;
+        }
+
+        static IAWSResource? SelectParentAWSResource(IResource resource) => resource switch
+        {
+            IAWSResource ar => ar,
+            IResourceWithParent rp => SelectParentAWSResource(rp.Parent),
+            _ => null
+        };
+        // parent -> children lookup
+        var parentChildLookup = appModel.Resources.OfType<IResourceWithParent>()
+            .Select(x => (Child: x, Root: SelectParentAWSResource(x.Parent)))
+            .Where(x => x.Root is not null)
+            .ToLookup(x => x.Root, x => x.Child);
+
+        // Sets the state of the resource and all of its children
+        async Task UpdateStateAsync(IAWSResource resource, Func<CustomResourceSnapshot, CustomResourceSnapshot> stateFactory)
+        {
+            await notificationService.PublishUpdateAsync(resource, stateFactory).ConfigureAwait(false);
+            foreach (var child in parentChildLookup[resource])
+            {
+                await notificationService.PublishUpdateAsync(child, stateFactory).ConfigureAwait(false);
+            }
+        }
+        // After the resource is provisioned, set its state
+        async Task AfterProvisionAsync(IAWSResource resource)
```

Comment:
I'm finding this class getting more complicated then it needs to be. For example the logic of this `AfterProvisionAsync` could just be moved into the `ProcessResourceAsync` method and get rid of this extra fire and forget `Task`. Especially if you make `UpdateStateAsync` a private method that all of the methods can call.

All I see that `Before StartAsync` needs to do is call the `ProvisionAWSResources` and fire and forget the return `Task`. The individual call for `ProcessResourceAsync` for the resource can take care of updating the starting and finishing state. Now I see as `BeforeStartAsync` is managing state with all of the complications that involves while `ProvisionAWSResources` has to take care of managing the task completion source.

Ultimately I want the code to be simpler as I feel this is getting overcomplicated and harder for others to quickly see what is happening.

Reply:
I'll do another iteration to simplify the provisioner. Getting away from the local methods will clean up most of it.